README.md
# Datatlas: Data Atlas

Data profiling and data quality checks


## Installation & Updates

Run the following to install:
```bash
pip install datatlas
```

Run one of the following to update:
```bash
pip install datatlas --upgrade

pip install datatlas -U
```


## Usage

```python
# Import the library
import datatlas as dat

# Print 'Hello Atlas!' to test if the package installation and import were successful
dat.say_hello('Atlas') 

# Generate an output df with detailed stats about the input df including warnings for data quality issues that might need to be addressed before model training
df_profile = dat.df_profiling(df) 

'''
df_profiling(df, nulls_threshold=50.0, zeroes_threshold=50.0, cardinality_threshold=50.0)
    Input(s):
        1) df: Dataframe to analyse
        2) nulls_threshold (int): Default is 50. Adds warning if >50% nulls detected
        3) zeroes_threshold (int): Default 50. Adds warning if >50% zeroes detected for numerical dtype columns
        4) cardinality_threshold (int): Default 50. Adds warning if >50% unique values for categorical dtype columns
    Output(s):
        1) Dataframe with statistics and data quality warnings about the input df
'''

```

## Versions
Please see the release_notes.txt file

## Contributing
[section under construction]




release_notes.txt

##### Datatlas Release Notes #####

1.0.4 @ 2020-11-14
* Added functionality that checks for date-like patterns and yes/no flag-like patterns

1.0.3 @ 2019-10-31
* Added functionality that predicts missing numerical data values using SGDRegressor - as an alternative to replacing missing values with a mean/max/etc

1.0.1 @ 2019-10-26
* Added more detailed dtype profiling using regex
* Refactored some of the code to make it more modular & easier for future working

1.0.0 @ 2019-09-23
* Function that takes a df input and outputs a df with a basic data profile & warnings

Note: wherever I miss out versions, it's probably because I messed up the Git or Python package distribution (I'll never understand wheels and eggs haha - if you get them, please teach me)


datatlas/datatlas_eda.py

"""
Datatlas: Data Atlas
* Data profiling

Author: Mei Yong
Updated date: 2019-09-21


	WIP Ideas
    * Distinct values
	* YesNo
	* Categorical
	* Mixed types based on regex
	* ID check
		- sequential score
		- Check for repeated value score - mask then group then average where zero is means more repeated values and one means less repeated values and closer to zero is more likely to be an ID
		-Check for Benford's law - basically if the values start with the same number or set of numbers
     * y/n check
        - Convert to string (keep the NaNs)
        - Check for common y/n values
	* Date check
        - Remove NaNs
        - Convert to string
        - Check regex and label
        - Count up the labels
        - If >60% date then date warning
	* Treat dates
		- split into year, month, day
		- days between current date and column date
		- convert to weekday, weekend, holidays
	* If high cardinality, check if can bucket smaller cats into other
	* Skew & should be log. must be positive. can add 1 to data to positive
	* If column should be split/pivoted
	* Only few different numbers which are actually categories
	* Treatments for time series data


"""

# Simple function for testing package installation and function calling
def say_hello(name):
	print(f"Hello {name}")
	
################################################################################


# Function that profiles a df
def df_profiling(df, nulls_threshold=50.0, zeroes_threshold=50.0, cardinality_threshold=50.0):
    '''
    Input(s):
        1) df - dataframe to analyse
        2) nulls_threshold - int - default 50. Adds warning if >50% nulls
        3) zeroes_threshold - int - default 50. Adds warning if >50% zeroes for numerical dtype columns
        4) cardinality_threshold - int - default 50. Adds warning if >50% nulls for categorical dtype columns
    Output(s):
        1) Dataframe with statistics and data quality warnings about the input df
    '''

    import pandas as pd
    import numpy as np
    
    rowcount = df.shape[0]
    
    # Datatype
    dtype_dict = {column : df[column].dtype.name for column in df}
    
    # Number of not-nulls
    count_dict = {column : df[column].count() for column in df}
    
    # Number of nulls
    null_dict_n = {column : df[column].isnull().sum() for column in df}
    
    # Percentage of nulls
    null_dict_p = {column : round(df[column].isnull().sum() / rowcount * 100, 2) for column in df}
    
    # Number of distinct categorical values
    dictinct_val_dict = {column: df[column].value_counts().count() \
                         for column in df if df[column].dtype in ['object'] }
    
    # Percentage of distinct categorical values
    cardinality_dict = {column: round(df[column].value_counts().count() / rowcount * 100 , 2) \
                         for column in df if df[column].dtype in ['object'] }
    
    # Number of zeroes
    zeroes_dict_n = {column : (df[column]==0).sum() \
                         for column in df if df[column].dtype in ['int64','float64'] }
    # Percentage of zeroes
    zeroes_dict_p = {column : round((df[column]==0).sum() / rowcount * 100, 2) \
                         for column in df if df[column].dtype in ['int64','float64'] }
    
    # Mean numerical value
    mean_dict = {column : df[column].mean() \
                         for column in df if df[column].dtype in ['int64','float64'] }
    
    # Min numerical value
    min_dict = {column : df[column].min() \
                         for column in df if df[column].dtype in ['int64','float64'] }
    
    # Quartile 1 - 25%
    q1_dict = {column : np.percentile(df[column], 25) \
                         for column in df if df[column].dtype in ['int64','float64'] }
    # Quartile 2 - 50%
    q2_dict = {column : np.percentile(df[column], 50) \
                         for column in df if df[column].dtype in ['int64','float64'] }
    # Quartile 3 - 75%
    q3_dict = {column : np.percentile(df[column], 75) \
                         for column in df if df[column].dtype in ['int64','float64'] }
    
    # Max numerical value
    max_dict = {column : df[column].max() \
                         for column in df if df[column].dtype in ['int64','float64'] }
    
    # Outliers
    
    
    # DQ Warnings
    warning_null_dict = {key: 1 for key,value in null_dict_p.items() if value >= nulls_threshold }
    warning_zeroes_dict = {key: 1 for key,value in zeroes_dict_p.items() if value >= zeroes_threshold }
    warning_cardinality_dict = {key: 1 for key,value in cardinality_dict.items() if value >= cardinality_threshold }
    warning_one_distinct_dict = {key: 1 for key,value in dictinct_val_dict.items() if value == 1 }
    
    warning_df = pd.DataFrame([
                                warning_null_dict
                                ,warning_zeroes_dict
                                ,warning_cardinality_dict
                                ,warning_one_distinct_dict
                                ]
                                , index = ['Nulls','Zeroes','Cardinality','Only_1_distinct']   
                                )
    
    final_warnings_list = []
    for column in warning_df:
        column_warnings = []
        if warning_df.loc['Nulls' , column] == 1:
            column_warnings.append('High_null_percentage')
        if warning_df.loc['Zeroes' , column] == 1:
            column_warnings.append('High_zeroes_percentage')
        if warning_df.loc['Cardinality' , column] == 1:
            column_warnings.append('High_cardinality')
        if warning_df.loc['Only_1_distinct' , column] == 1:
            column_warnings.append('Only_1_unique_value')
        final_warnings_list.append(column_warnings)
        
    final_warnings_dict = dict(zip(warning_df.columns , final_warnings_list))
            
            
    
    df_profile = pd.DataFrame([
                        dtype_dict
                        ,count_dict
                        ,null_dict_n
                        ,null_dict_p
                        ,dictinct_val_dict
                        ,cardinality_dict
                        ,zeroes_dict_n
                        ,zeroes_dict_p
                        ,mean_dict
                        ,min_dict
                        ,q1_dict
                        ,q2_dict
                        ,q3_dict
                        ,max_dict
                        #outliers
                        ,final_warnings_dict
                        ]
                        , index = ['Dtype','Count','Null(n)','Null(p)','Distinct_Values','Cardinality',
                                   'Zeroes(n)','Zeroes(p)','Mean','Min','Q1','Q2','Q3','Max','DQ_Warnings']       
                        )
    
    return df_profile


###################################################################################################

### To enable this module to be called from another Python file
    
if __name__ == '__main__':
    say_hello()
    df_profiling()




datatlas/datatlas_missing_values.py

"""
Datatlas: Data Atlas
* Predicting missing values

Author: Mei Yong
Updated date: 2019-10-30

WIP Ideas
    * numerical values - done
    * categorical values - not started
    * hyperparameter tuning - but that would take awhile so maybe only run it if the optional is switched on?

"""

# Import libraries
import pandas as pd
import numpy as np


# Helper function for one-hot encoding of categorical features
def one_hot_encode(df):
    # Identify only categorical features
    categorical_features = df.select_dtypes(include='object').columns.tolist()
    # Loop through categorical features
    for col in categorical_features:
        dummies = pd.get_dummies(df[col], prefix=col, dummy_na=False) # Create dummy features
        df = df.drop(col,1) # Drop the original features
        df = pd.concat([df, dummies], axis=1) # Add the dummy features to the input df to replace the original features
    return df
    

# Main function for imputing missing values
def predict_missing_values(df, impute_column, datatype, impute_model=True, impute_stats=True, console_messages=True):
    '''
    Input(s):
        1) df: Dataframe of all columns including the column with missing values
        2) impute_column (string): Name of the column containing missing values that you want to predict and impute using machine learning
        3) datatype (string): Choose one of the following - numeric, categoric
        4) impute_model (boolean): Default True. See outputs docstring below for more info
        4) stats (boolean): Default True. See outputs docstring below for more info
        5) console_messages (boolean): Default True. Outputs status messages in the console to indicate progress of the model training and predictions
    Output(s):
        1) Dataframe where the nulls in the column to impute are replaced with their predicted values
        2) impute_model (Sklearn estimator object) (optional): The trained model that was used to predict the missing values that can be later used to predict missing values of new live data
        3) impute_stats (dictionary) (optional): A dictionary containing stats about the impute column before and after the predictions. Also includes rmse (for numeric) or accuracy (for categoric) model performance evaluation metrics
    '''
    
    # Keep a copy of the original df to be returned at the end with imputed values
    original_df = df.copy()
    
    # Cleanse the train set - dropping high nulls, zeroes, cardinality
    rowcount = df.shape[0]
    null_profile = {column : round(df[column].isnull().sum() / rowcount * 100, 2) for column in df}
    zeroes_profile = {column : round((df[column]==0).sum() / rowcount * 100, 2) \
                         for column in df if df[column].dtype in ['int64','float64'] }
    cardinality_profile = {column: round(df[column].value_counts().count() / rowcount * 100 , 2) \
                         for column in df if df[column].dtype in ['object'] }
    dq_check = pd.DataFrame([null_profile, zeroes_profile, cardinality_profile]\
                            ,index=['nulls','zeroes','cardinality'])
    dq_check = dq_check.fillna(0)
    del rowcount, null_profile, zeroes_profile, cardinality_profile
    
    del_cols = []
    for col in dq_check:
        if dq_check.loc['nulls',col]>25 \
        or dq_check.loc['zeroes',col]>25 \
        or dq_check.loc['cardinality',col]>50:
            del_cols.append(col)
    
    # If the column to be imputed is in the list of columns to drop, remove it
    if impute_column in del_cols:
        del_cols.remove(impute_column)
    
    # Drop all of the bad quality columns
    df = df.drop(del_cols, axis=1)
    
    # Drop rows if any columns other than the target one has nulls
    target_col = df.pop(impute_column)
    df = df.dropna(axis=0, how='any')
    df = df.merge(target_col, left_index=True, right_index=True)
    
    # One hot encode the categorical columns
    target_col = df.pop(impute_column) # pop out the impute column
    df = one_hot_encode(df) # perform one hot encoding
    df[impute_column] = target_col # stick the impute column back on


    if datatype == 'numeric':
        
        # Replace all zeroes in the target column with nulls
        df[impute_column] = df[impute_column].replace(0, np.nan)
        
        # Calculate the percentage of nulls in the column to be imputed and print in console
        null_n = sum(df[impute_column].isnull())
        null_p = sum(df[impute_column].isnull()) / df[impute_column].shape[0] * 100
        
        if console_messages == True:
            print(f"Datatlas: {round(null_p,2)}% of the target column is null and will be predicted")
        
        # Split the set into train where the impute column is not null and pred where it is null
        train = df[ df[impute_column].isnull()==False ]
        pred = df[ df[impute_column].isnull()==True ]
        
        # Split the train and pred sets into X and y
        y_train, X_train = train.pop(impute_column), train.copy()
        y_pred, X_pred = pred.pop(impute_column), pred.copy()
        
        # Train a regression model using the train set
        if console_messages == True:
            print("Datatlas: Training a Stochastic Gradient Descent regression model to predict the missing numerical values")
               
        from sklearn.linear_model import SGDRegressor
        sgd_reg = SGDRegressor(max_iter=1000, tol=1e-3, penalty=None, eta0=0.1)
        sgd_reg.fit(X_train, y_train)
        
        # Get predictions using the trained model
        if console_messages == True:
            print("Datatlas: Predicting the missing numerical values")
        y_pred = sgd_reg.predict(X_pred)
        y_pred = pd.Series(y_pred, index=X_pred.index)
        
        # Replace the nulls and zeroes in the original df with the imputed/predicted values
        original_df[impute_column] = original_df[impute_column].replace(0, np.nan)
        original_df[impute_column] = original_df[impute_column].fillna(value=y_pred)
        
        
        # Different objects get returned based on the parameters inputted into the function
        
        if impute_stats == False:
            
            if impute_model == False:
                return original_df # No stats, no model
            elif impute_model == True:
                return original_df, sgd_reg # No stats, yes model
        
        # If the impute stats option is turned on, perform calculations for the dictionary and then return different objects depending on whether optional trained impute model is also turned on
        if impute_stats == True:
            
            # Get performance stats using cross val on the train set
            if console_messages == True:
                print("Datatlas: Getting model performance stats. Please be patient - this step uses cross validation with 3 folds")
            from sklearn.model_selection import cross_val_predict
            from sklearn.metrics import mean_squared_error
            from math import sqrt
            y_train_pred = cross_val_predict(sgd_reg, X_train, y_train, cv=3)
            
            impute_stats_dict = {'imputed_column': impute_column
                                ,'rmse': sqrt(mean_squared_error(y_train, y_train_pred))
                                ,'min_value': min(df[impute_column].dropna())
                                ,'max_value': max(df[impute_column].dropna())
                                ,'mean_value': df[impute_column].mean()
                                ,'original_nulls_num': null_n
                                ,'original_nulls_p': null_p
                                ,'remaining_nulls_num': sum(original_df[impute_column].isnull())
                                ,'remaining_nulls_p': sum(original_df[impute_column].isnull()) / original_df[impute_column].shape[0] * 100
                        }
            
            if impute_model == False:
                return original_df, impute_stats_dict # Yes stats, no model
            elif impute_model == True:
                return original_df, sgd_reg, impute_stats_dict # Yes stats, yes model
        
        
    if datatype == 'categoric':
        import sys
        sys.exit("Datatlas: Sorry, functionality to predict categoric values is still a work in progress")


"""
# Test
df = pd.read_csv(r"C:\Users\shaom\Desktop\Machine_Learning\projects\data\airbnb\listings.csv")
new_df_sgd, sgd_model, impute_stats_sgd = predict_missing_values(df, impute_column='cleaning fee', datatype='numeric', impute_model=True, impute_stats=True, console_messages=False)
"""


###################################################################################################

### To enable this module to be called from another Python file
    
if __name__ == '__main__':
    predict_missing_values()






